File: forms/sundowner-syndrome-assessment/front-end-form-with-html/js/sundowner-grader.js
// Sundowner syndrome grader. Pure functions: take an `AssessmentData`
// object, return the CMAI total, NPI total, severity band, and a list of
// fired rules summarising which scoring bands and elevated NPI domains
// were triggered.
//
// CMAI severity thresholds (from the AGENTS.md spec):
//   29-45    -> mild       (occasional restlessness, redirectable)
//   46-75    -> moderate   (daily episodes, requires intervention)
//   76-120   -> severe     (aggressive behaviour, safety risk)
//   >120     -> critical   (self-harm risk, requires constant supervision)
//
// CMAI score is the sum of all 29 items each in 1..7 (range 29..203).
// Items not yet answered (score 0) are scored as 1 ("Never") for the
// purposes of computing the total, which keeps a partially-complete form
// classifiable as "mild" rather than yielding a misleading low total.
// We separately report `cmaiAnsweredCount` so the report can show
// completion progress.
//
// NPI total is the sum across all 12 domains of (frequency * severity),
// each in [0, 12], range 0..144.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').Severity} Severity
 * @typedef {import('./types.js').FiredRule} FiredRule
 */

(function () {
'use strict';
window.SundownerSyndromeAssessment = window.SundownerSyndromeAssessment || {};
const NS = window.SundownerSyndromeAssessment;
const {
  severityFromCMAI,
  severityLabel,
  cmaiItems,
  npiDomains,
  CMAI_ITEM_IDS,
  NPI_DOMAIN_KEYS
} = NS;

/**
 * Sum the 29 CMAI items. Unanswered items (stored as 0) are treated as
 * 1 ("Never") so a partly-complete form still classifies sensibly.
 * @param {AssessmentData} data
 * @returns {{ total: number, answered: number }}
 */
function sumCMAI(data) {
  const cmai = data.behaviouralSymptoms?.cmai || {};
  let total = 0;
  let answered = 0;
  for (const id of CMAI_ITEM_IDS) {
    const v = cmai[id];
    if (typeof v === 'number' && v >= 1 && v <= 7) {
      total += v;
      answered++;
    } else {
      // Unanswered: score as 1 (Never)
      total += 1;
    }
  }
  return { total, answered };
}

/**
 * Sum the 12 NPI domain scores (frequency * severity).
 * @param {AssessmentData} data
 * @returns {{ total: number, answered: number, perDomain: { key: string, label: string, score: number, frequency: number, severity: number }[] }}
 */
function sumNPI(data) {
  const npi = data.behaviouralSymptoms?.npi || {};
  let total = 0;
  let answered = 0;
  /** @type {{ key: string, label: string, score: number, frequency: number, severity: number }[]} */
  const perDomain = [];
  for (const domain of npiDomains) {
    const entry = npi[domain.key] || { frequency: 0, severity: 0 };
    const f = Number(entry.frequency) || 0;
    const s = Number(entry.severity) || 0;
    const score = (f >= 1 && f <= 4 && s >= 1 && s <= 3) ? f * s : 0;
    if (score > 0) answered++;
    total += score;
    perDomain.push({
      key: domain.key,
      label: domain.label,
      score,
      frequency: f,
      severity: s
    });
  }
  return { total, answered, perDomain };
}

/**
 * Grade the sundowner assessment.
 *
 * @param {AssessmentData} data
 * @returns {{
 *   cmaiScore: number,
 *   npiScore: number,
 *   severity: Severity,
 *   cmaiAnsweredCount: number,
 *   npiAnsweredCount: number,
 *   firedRules: FiredRule[]
 * }}
 */
function gradeSundowner(data) {
  const cmaiSum = sumCMAI(data);
  const npiSum = sumNPI(data);
  const cmaiScore = cmaiSum.total;
  const npiScore = npiSum.total;
  const severity = severityFromCMAI(cmaiScore);

  /** @type {FiredRule[]} */
  const firedRules = [];

  // ─── CMAI band rule ───────────────────────────────────────────
  firedRules.push({
    id: 'CMAI-BAND',
    category: 'CMAI Total',
    description: `CMAI total ${cmaiScore} of 203 (${cmaiSum.answered} of 29 items observed).`,
    detail: `Severity band: ${severityLabel(severity)} (CMAI thresholds: 29-45 mild, 46-75 moderate, 76-120 severe, >120 critical).`
  });

  // ─── Highly-elevated CMAI items (>= 5: at least daily) ────────
  const highCmaiItems = cmaiItems.filter((item) => {
    const v = data.behaviouralSymptoms?.cmai?.[item.id];
    return typeof v === 'number' && v >= 5;
  });
  if (highCmaiItems.length > 0) {
    firedRules.push({
      id: 'CMAI-DAILY',
      category: 'CMAI Items',
      description: `${highCmaiItems.length} CMAI item(s) occurring at least daily.`,
      detail: highCmaiItems.map((i) => `#${i.number} ${i.label}`).join('; ')
    });
  }

  // ─── NPI total band ──────────────────────────────────────────
  const npiBand = npiScore >= 48 ? 'markedly elevated'
    : npiScore >= 24 ? 'moderately elevated'
    : npiScore >= 12 ? 'mildly elevated'
    : 'within normal limits';
  firedRules.push({
    id: 'NPI-TOTAL',
    category: 'NPI Total',
    description: `NPI total ${npiScore} of 144 across 12 domains (${npiSum.answered} domain(s) endorsed).`,
    detail: `NPI total is ${npiBand}.`
  });

  // ─── Elevated NPI sub-domains (score >= 4) ───────────────────
  const elevatedDomains = npiSum.perDomain.filter((d) => d.score >= 4);
  if (elevatedDomains.length > 0) {
    firedRules.push({
      id: 'NPI-DOMAINS',
      category: 'NPI Domains',
      description: `${elevatedDomains.length} NPI domain(s) with score >= 4.`,
      detail: elevatedDomains
        .map((d) => `${d.label} (F${d.frequency} x S${d.severity} = ${d.score})`)
        .join('; ')
    });
  }

  return {
    cmaiScore,
    npiScore,
    severity,
    cmaiAnsweredCount: cmaiSum.answered,
    npiAnsweredCount: npiSum.answered,
    firedRules
  };
}

Object.assign(window.SundownerSyndromeAssessment, {
  gradeSundowner,
  sumCMAI,
  sumNPI
});
})();

File: forms/sundowner-syndrome-assessment/back-end-with-loco/src/engine/sundowner_grader.rs
//! Sundowner grader module.

// Sundowner syndrome grader. Pure functions: take an `AssessmentData`
// object, return the CMAI total, NPI total, severity band, and a list
// of fired rules summarising which scoring bands and elevated NPI
// domains were triggered.
//
// Unanswered CMAI items (score 0) are scored as 1 ("Never") so that a
// partially-complete form still classifies sensibly. `cmai_answered_count`
// is reported separately so the report can show completion progress.

use super::flagged_issues::detect_additional_flags;
use super::sundowner_rules::{cmai_items, npi_domains};
use super::types::{AssessmentData, FiredRule, GradingResult};
use super::utils::{severity_from_cmai, severity_label};

/// Sum of the 29 CMAI items. Unanswered (0) items are treated as 1 (Never).
fn sum_cmai(data: &AssessmentData) -> (i32, u32) {
    let mut total = 0i32;
    let mut answered = 0u32;
    for item in cmai_items() {
        let v = data
            .behavioural_symptoms
            .cmai
            .get(item.id)
            .copied()
            .unwrap_or(0);
        if (1..=7).contains(&v) {
            total += v;
            answered += 1;
        } else {
            total += 1;
        }
    }
    (total, answered)
}

/// Per-domain NPI score (frequency * severity, 0 if either is out of range).
pub struct NpiDomainSum {
    /// Key.
    pub key: &'static str,
    /// Label.
    pub label: &'static str,
    /// Score.
    pub score: i32,
    /// Frequency.
    pub frequency: i32,
    /// Severity.
    pub severity: i32,
}

/// Sum of the 12 NPI domain scores.
fn sum_npi(data: &AssessmentData) -> (i32, u32, Vec<NpiDomainSum>) {
    let mut total = 0i32;
    let mut answered = 0u32;
    let mut per_domain: Vec<NpiDomainSum> = Vec::new();
    for domain in npi_domains() {
        let entry = data.behavioural_symptoms.npi.get(domain.key);
        let f = entry.map(|e| e.frequency).unwrap_or(0);
        let s = entry.map(|e| e.severity).unwrap_or(0);
        let score = if (1..=4).contains(&f) && (1..=3).contains(&s) {
            f * s
        } else {
            0
        };
        if score > 0 {
            answered += 1;
        }
        total += score;
        per_domain.push(NpiDomainSum {
            key: domain.key,
            label: domain.label,
            score,
            frequency: f,
            severity: s,
        });
    }
    (total, answered, per_domain)
}

/// Grade the sundowner assessment.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let (cmai_score, cmai_answered) = sum_cmai(data);
    let (npi_score, npi_answered, per_domain) = sum_npi(data);
    let severity = severity_from_cmai(cmai_score);

    let mut fired_rules: Vec<FiredRule> = Vec::new();

    // ─── CMAI-BAND ──────────────────────────────────────────────
    fired_rules.push(FiredRule {
        id: "CMAI-BAND".to_string(),
        category: "CMAI Total".to_string(),
        description: format!(
            "CMAI total {cmai_score} of 203 ({cmai_answered} of 29 items observed)."
        ),
        detail: format!(
            "Severity band: {} (CMAI thresholds: 29-45 mild, 46-75 moderate, 76-120 severe, >120 critical).",
            severity_label(&severity)
        ),
    });

    // ─── CMAI-DAILY (items rated >= 5) ─────────────────────────
    let high_cmai: Vec<&'static super::sundowner_rules::CmaiItem> = cmai_items()
        .iter()
        .filter(|item| {
            data.behavioural_symptoms
                .cmai
                .get(item.id)
                .copied()
                .unwrap_or(0)
                >= 5
        })
        .collect();
    if !high_cmai.is_empty() {
        let detail = high_cmai
            .iter()
            .map(|i| format!("#{} {}", i.number, i.label))
            .collect::<Vec<_>>()
            .join("; ");
        fired_rules.push(FiredRule {
            id: "CMAI-DAILY".to_string(),
            category: "CMAI Items".to_string(),
            description: format!(
                "{} CMAI item(s) occurring at least daily.",
                high_cmai.len()
            ),
            detail,
        });
    }

    // ─── NPI-TOTAL band ────────────────────────────────────────
    let npi_band = if npi_score >= 48 {
        "markedly elevated"
    } else if npi_score >= 24 {
        "moderately elevated"
    } else if npi_score >= 12 {
        "mildly elevated"
    } else {
        "within normal limits"
    };
    fired_rules.push(FiredRule {
        id: "NPI-TOTAL".to_string(),
        category: "NPI Total".to_string(),
        description: format!(
            "NPI total {npi_score} of 144 across 12 domains ({npi_answered} domain(s) endorsed)."
        ),
        detail: format!("NPI total is {npi_band}."),
    });

    // ─── NPI-DOMAINS (score >= 4) ──────────────────────────────
    let elevated: Vec<&NpiDomainSum> = per_domain.iter().filter(|d| d.score >= 4).collect();
    if !elevated.is_empty() {
        let detail = elevated
            .iter()
            .map(|d| format!("{} (F{} x S{} = {})", d.label, d.frequency, d.severity, d.score))
            .collect::<Vec<_>>()
            .join("; ");
        fired_rules.push(FiredRule {
            id: "NPI-DOMAINS".to_string(),
            category: "NPI Domains".to_string(),
            description: format!("{} NPI domain(s) with score >= 4.", elevated.len()),
            detail,
        });
    }

    let additional_flags = detect_additional_flags(data);
    let timestamp = chrono::Utc::now().to_rfc3339();

    GradingResult {
        cmai_score,
        npi_score,
        severity,
        cmai_answered_count: cmai_answered,
        npi_answered_count: npi_answered,
        fired_rules,
        additional_flags,
        timestamp,
    }
}
